trinity/utils/datastructures.py

```diff
@@ -286,3 +297,315 @@ def complete(self, batch_id: int, completed: Tuple[TTask, ...]) -> None:
     def __contains__(self, task: TTask) -> bool:
         """Determine if a task has been added and not yet completed"""
         return task in self._tasks
+
+
+class BaseTaskPrerequisites(Generic[TTask, TSubtask]):
+    """
+    Keep track of which prerequisites on a task are complete. It is used internally by
+    :class:`OrderedTaskPreparation`
+    """
+    _subtasks: Iterable[TSubtask]
+    _completed_subtasks: Set[TSubtask]
+    _task: TTask
+
+    @classmethod
+    def from_enum(cls, subtasks: Type[TSubtask]) -> 'Type[BaseTaskPrerequisites[Any, Any]]':
+
+        if len(subtasks) < 1:
+            raise ValidationError("There must be at least one subtask to track completions")
+
+        return type('CompletionFor' + subtasks.__name__, (cls, ), dict(_subtasks=subtasks))
```

Comment:
Alright, I think I found one source of confusion for me.  The term subtask here doesn't refer to actual [`asyncio.Task`] objects does it.  Rather, it's just something to represent the "jobs" that need to be done before the master task can be done.  Thoughts on using "job"?

Reply:
Oh yeah, this name is a bug (stayed the old name). Every name subtask should have been changed to prerequisite.